src/EditorFeatures/CSharp/EventHookup/EventHookupCommandHandler_TabKeyCommand.cs

```diff
@@ -85,116 +78,179 @@ private void HandleTabWorker(ITextView textView, ITextBuffer subjectBuffer, Acti
+    private async Task ExecuteCommandAsync(
+        TabKeyCommandArgs args,
+        Action nextHandler,
+        SnapshotSpan applicableToSpan,
+        Document document,
+        Task<string> getEventNameTask,
+        SnapshotPoint initialCaretPoint)
+    {
+        try
         {
-            eventHandlerMethodName = EventHookupSessionManager.CurrentSession.GetEventNameTask.WaitAndGetResult(cancellationToken);
+            _threadingContext.ThrowIfNotOnUIThread();
+            await ExecuteCommandWorkerAsync().ConfigureAwait(false);
         }
-
-        if (eventHandlerMethodName == null ||
-            EventHookupSessionManager.CurrentSession.TextView != textView)
+        catch (OperationCanceledException)
+        {
+        }
+        catch (Exception ex) when (FatalError.ReportAndCatch(ex))
         {
-            nextHandler();
-            EventHookupSessionManager.CancelAndDismissExistingSessions();
-            return;
         }
 
-        // This tab means we should generate the event handler method. Begin the code
-        // generation process.
-        GenerateAndAddEventHandler(textView, subjectBuffer, eventHandlerMethodName, nextHandler, cancellationToken);
-    }
-
-    private void GenerateAndAddEventHandler(ITextView textView, ITextBuffer subjectBuffer, string eventHandlerMethodName, Action nextHandler, CancellationToken cancellationToken)
-    {
-        _threadingContext.ThrowIfNotOnUIThread();
+        return;
 
-        using (Logger.LogBlock(FunctionId.EventHookup_Generate_Handler, cancellationToken))
+        async Task ExecuteCommandWorkerAsync()
         {
-            EventHookupSessionManager.CancelAndDismissExistingSessions();
+            var textView = args.TextView;
+            var subjectBuffer = args.SubjectBuffer;
+
+            var factory = document.Project.Solution.Workspace.Services.GetRequiredService<IBackgroundWorkIndicatorFactory>();
+            using var waitContext = factory.Create(
+                textView,
+                applicableToSpan,
+                CSharpEditorResources.Generating_event);
+
+            var cancellationToken = waitContext.UserCancellationToken;
+
+            var eventHandlerMethodName = await getEventNameTask.WithCancellation(cancellationToken).ConfigureAwait(false);
 
-            var workspace = textView.TextSnapshot.TextBuffer.GetWorkspace();
-            if (workspace == null)
+            var solutionAndRenameSpan = await TryGetNewSolutionWithAddedMethodAsync(
+                document, eventHandlerMethodName, initialCaretPoint.Position, cancellationToken).ConfigureAwait(false);
+
+            // We're about to make an edit ourselves.  so disable the cancellation that happens on editing.
+            waitContext.CancelOnEdit = false;
+
+            // switch back to the UI thread.
+            await _threadingContext.JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
+            _threadingContext.ThrowIfNotOnUIThread();
```

Comment:
💡 This assertion is 100% redundant. `SwitchToMainThreadAsync` throws if the switch did not occur.

Reply:
k